python/lsst/analysis/tools/actions/vector/calcRhoStatistics.py

```diff
@@ -0,0 +1,649 @@
+from .ellipticity import CalcE, CalcEDiff
+from .vectorActions import FractionalDifference
+
+if TYPE_CHECKING:
+    from treecorr import GGCorrelation, KKCorrelation
+
+    from ...interfaces import KeyedDataSchema
+
+_LOG = logging.getLogger(__name__)
+
+
+class BinnedCorr2Config(Config):
+    """A Config class that holds the various parameters supported by treecorr.
+
+    The fields in this class correspond to the parameters that can be passed to
+    BinnedCorr2 in `treecorr`, which is the base class for all two-point
+    correlation function calculations. The default values set for the fields
+    are identical to the default values set in v4.2 of `treecorr`. The
+    parameters that are excluded in this class are
+    'verbose', 'log_file', 'output_dots', 'rng' and 'pairwise' (deprecated).
+    For details about these options, see the documentation for `treecorr`:
+    https://rmjarvis.github.io/TreeCorr/_build/html/correlation2.html
+
+    A separate config class is used instead
+    of constructing a `~lsst.pex.config.DictField` so that mixed types can be
+    supported and the config can be validated at the beginning of the
+    execution. The ``toDict`` provides a minimal dictionary that override only
+    the default values and excludes the key-values pairs when the item is None.
+
+    Notes
+    -----
+    This is intended to be used in CalcRhoStatistics class.
+    """
+
+    nbins = Field[int](
+        doc=(
+            "How many bins to use. "
+            "(Exactly three of nbins, bin_size, min_sep, max_sep "
+            "are required. If nbins is not given, it will be "
+            "calculated from the values of the other three, "
+            "rounding up to the next highest integer. "
+            "In this case, bin_size will be readjusted to account "
+            "for this rounding up."
+        ),
+        optional=True,
+        check=lambda x: x > 0,
+    )
+
+    bin_size = Field[float](
+        doc=(
+            "The width of the bins in log(separation). "
+            "Exactly three of nbins, bin_size, min_sep, max_sep are required. "
+            "If bin_size is not given, it will be calculated from the values "
+            "of the other three."
+        ),
+        optional=True,
+    )
+
+    min_sep = Field[float](
+        doc=(
+            "The minimum separation in units of sep_units, if relevant. "
+            "Exactly three of nbins, bin_size, min_sep, max_sep are required. "
+            "If min_sep is not given, it will be calculated from the values "
+            "of the other three."
+        ),
+        optional=True,
+    )
+
+    max_sep = Field[float](
+        doc=(
+            "The maximum separation in units of sep_units, if relevant. "
+            "Exactly three of nbins, bin_size, min_sep, max_sep are required. "
+            "If max_sep is not given, it will be calculated from the values "
+            "of the other three."
+        ),
+        optional=True,
+    )
```

Comment:
What happens if three of them aren't set?

Reply:
Assuming it gets through to TreeCorr, then TreeCorr would raise an exception in that case.  Not sure if that's what you want here or not.

Reply:
But it wouldn't get to `treecorr` because an exception would be raised here: https://github.com/lsst/analysis_tools/blob/041b14cb1ecba793d69d79cca0623ef01af8db96/python/lsst/analysis/tools/actions/vector/calcRhoStatistics.py#L323-L330

This duplicates what `treecorr` does, but that is desirable because we want misconfiguration to be captured at config time, not at runtime, which is what would happen if we let `treecorr` handle it.

Reply:
I saw the catch below, makes sense now.